docs/proposals/cross-mesh-ux.md

```diff
@@ -0,0 +1,229 @@
+
+This proposal allows the operators of one `Mesh` to expose services to other `Meshes`
+via `MeshGateway`, the same interface for exposing to external traffic.
+
+The proposal addresses UX both for _exposing_ and the _consuming_
+meshes and only HTTP/S services.
+
+## Design
+
+In the following examples, the mesh exposing its services will be `expose` and
+the mesh consuming another mesh's services will be `consume`.
+
+<details>
+<summary>Recap: <code>MeshGateway</code> configuration</summary>
+
+```
+type: MeshGateway
+mesh: expose
+name: edge-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8080
+    protocol: HTTP
+    hostname: foo.example.com
+    tags:
+      port: "8080"
+---
+type: MeshGatewayRoute
+mesh: expose
+name: edge-gateway-route
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+conf:
+  http:
+    rules:
+      - matches:
+          - path:
+              match: PREFIX
+              value: /api
+        backends:
+          - destination:
+              kuma.io/service: server
+```
+
+</details>
+
+### `expose` UX
+
+The main UX feature will be a mechanism to configure a
+`MeshGateway` to permit traffic from services in one
+`Mesh` to reach the services behind the `MeshGateway`.
+The configuration will make it possible to expose `MeshGatewayRoute`s depending
+on cross-mesh traffic's source `Mesh` and source _tags_.
+
+First of all, we require users to explicitly enable cross-mesh traffic on
+the `MeshGateway` listener as well as mTLS on any meshes doing cross-mesh
+traffic.
+This option will allow us to configure the corresponding Envoy listener
+with all mesh CAs.
+
+```
+type: MeshGateway
+mesh: expose
+name: edge-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8090
+    protocol: HTTPS
+    hostname: expose.mesh # optional
+    tags:
+      name: mesh-listener
+    tls:
+      mode: TERMINATE
+      options:
+        crossMesh: true # new option
+```
+
+There are two ways we can enable configuring more specific access rules.
+
+#### `MeshGatewayTrafficPermission`
+
+The first is to add a _new resource_ that defines which
+cross-mesh traffic is allowed for a matching `MeshGateway` listener and
+`routeName`.
+
+```
+type: MeshGatewayTrafficPermission
+mesh: expose
+name: cross-mesh
+selector: # these match a listener
+- tags:
+    kuma.io/service: edge-gateway
+    name: mesh-listener
+default: deny
+rules:
+- routeName: cross-mesh
+  action: allow
+  sources:
+  - mesh:
+      name: consume
+      tags:
+        kuma.io/service: client
+```
+
+This concentrates `MeshGateway` permission management into its own resource.
+This may be a more generally useful abstraction given potential
+other kinds of `sources`. For example, one for filtering based on JWT token:
+
+```
+  - routeName: cross-mesh
+    action: allow
+    sources:
+    - user:
+        email: abc@xyz.com
+```
+
+#### Inline `MeshGatewayRoute`
```

Comment:
I agree this option might look simpler but reduces the flexibility for this in the future.